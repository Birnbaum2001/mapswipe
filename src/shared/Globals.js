// @flow
/**
 * @author Pim de Witte (pimdewitte.me/pimdewitte95@gmail.com). Copyright MSF UK 2016.
 *
 * Globals is a class that stores static variables throughout the application.
 * This include things such as the AuthenticationManager, the Model, and the Database connection.
 * It also stores stuff such as the screen dimentions and server variables.
 */

import { Platform, Dimensions } from 'react-native';

// const AuthManager = require('./AuthManager')

// var authManager = new AuthManager();

// var Analytics = require('react-native-firebase-analytics');

// const Database = require('./Database');
import Database from './Database';

// FIXME: check the old calculation to include status bar and soft menu
const screenHeight = Dimensions.get('window').height;

module.exports = {
    STORE_KEY: 'a56z0fzrNpl^2',
    BASE_URL: 'http://api.mapswipe.org:3000',
    FILE_PATH: '',
    TOP_OFFSET: Platform.OS === 'android' ? 0 : 20,
    SCREEN_WIDTH: Dimensions.get('window').width,
    SCREEN_HEIGHT: screenHeight,
    COLOR: {
        ORANGE: '#C50',
        DARKBLUE: '#0F3274',
        LIGHTBLUE: '#6EA8DA',
        DARKGRAY: '#999',
    },
    TASKS_PROCESSING: 0,
    TILE_VIEW_HEIGHT: (screenHeight - 40 - 30), // 40 is top bar, 30 is bottom bar
    TILES_PER_VIEW_X: 2,
    TILES_PER_VIEW_Y: 3,
    TILE_VIEW_WIDTH: 1,
    // AUTH_MANAGER: authManager,
    DB: Database,
    GRADIENT_COUNT: 0,
    TUT_LINK: 'http://www.missingmaps.org/blog/2016/07/18/mapswipetutorial/',
    // ANALYTICS: Analytics,
    SUPPORTED_PROJECT_TYPES: [1, 2],
};
