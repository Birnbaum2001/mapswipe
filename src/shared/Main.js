/**
 * @author Pim de Witte (pimdewitte.me/pimdewitte95@gmail.com). Copyright MSF UK 2016.
 *
 * Main is the main class that is called from both Android and iOS on application startup.
 * It initializes the application and controls which scene is rendered to the end user through
 * the Navigator component.
 */

// @flow

import * as React from 'react';
import {
    Image,
    SafeAreaView,
    StyleSheet,
    Text,
    View,
} from 'react-native';
import Button from 'apsl-react-native-button';
import { createStackNavigator } from 'react-navigation';
import Login from './views/Login';
import BuildingFootprintValidator from './views/BuildingFootprint';
import ChangeDetectionScreen from './views/ChangeDetection';
import Mapper from './views/Mapper';
import ProjectNav from './views/ProjectNav';
import WelcomeScreen from './views/Welcome';
import WebviewWindow from './views/WebviewWindow';
import {
    COLOR_DEEP_BLUE,
} from './constants';

const MessageBarAlert = require('react-native-message-bar').MessageBar;
const { MessageBarManager } = require('react-native-message-bar');
const Modal = require('react-native-modalbox');

const ProjectView = require('./views/ProjectView');
const GLOBAL = require('./Globals');


const style = StyleSheet.create({
    startButton: {
        backgroundColor: '#ff0000',
        alignItems: 'stretch',
        height: 50,
        padding: 12,
        borderRadius: 5,
        borderWidth: 0.1,
        position: 'absolute',
        bottom: 20,
        left: 20,
        width: 260,
    },
    safeArea: {
        flex: 1,
        backgroundColor: COLOR_DEEP_BLUE,
    },
    pic: {
        height: 150,
        width: 150,
    },
    mainContainer: {
        height: GLOBAL.SCREEN_HEIGHT,
        width: GLOBAL.SCREEN_WIDTH,
        flex: 1,
        marginTop: GLOBAL.TOP_OFFSET,
    },
    header: {
        fontWeight: '700',
        color: '#212121',
        fontSize: 18,
    },
    modal: {
        padding: 20,
    },
    modal3: {
        marginTop: 10,
        height: 300,
        width: 300,
        backgroundColor: '#ffffff',
        borderRadius: 2,
        alignItems: 'center',
    },
});

type State = {
    isDisabled: bool,
    level: number,
    levelObject: Object,
};

class Main extends React.Component<{}, State> {
    constructor(props: {}) {
        super(props);
        this.state = {
            isDisabled: false,
            level: GLOBAL.DB.getLevel(),
            levelObject: GLOBAL.DB.getLevelObject(),
        };
    }

    /**
     * Starts the level up timer and register the notification bar
     */
    componentDidMount() {
        const parent = this;
        // GLOBAL.ANALYTICS.logEvent('mapswipe_open');
        MessageBarManager.registerMessageBar(parent.alert);

        parent.checkInterval = setInterval(() => {
            if (GLOBAL.DB.getPendingLevelUp() > 0) {
                parent.openModal3(GLOBAL.DB.getPendingLevelUp());
                GLOBAL.DB.setPendingLevelUp(-1);
            }
        }, 500);
    }

    componentWillUnmount() {
        clearInterval(this.checkInterval);
    }

    alert: ?React.ComponentType<{}>;

    checkInterval: IntervalID;

    modal3: ?Modal;

    openModal3(level: number) {
        this.setState({
            levelObject: GLOBAL.DB.getCustomLevelObject(level),
            level,
        });
        if (this.modal3) {
            this.modal3.open();
        }
    }

    closeModal3() {
        // $FlowFixMe
        this.modal3.close();
        GLOBAL.DB.stopPopup();
    }

    render() {
        const { isDisabled, level, levelObject } = this.state;
        return (
        <SafeAreaView style={style.safeArea}>
            <View style={style.mainContainer}>
                <RootStack />
                <Modal
                    style={[style.modal, style.modal3]}
                    backdropType="blur"
                    position="center"
                    ref={(r) => { this.modal3 = r; }}
                    isDisabled={isDisabled}
                >
                    <Text style={style.header}>
                        {`You are now level ${level}`}
                    </Text>
                    <Image style={style.pic} key={level} source={levelObject.badge} />
                    <Button
                        style={style.startButton}
                        onPress={this.closeModal3}
                        textStyle={{ fontSize: 13, color: '#ffffff', fontWeight: '700' }}
                    >
                    Close
                    </Button>
                </Modal>
                <MessageBarAlert ref={(r) => { this.alert = r; }} />
            </View>
            </SafeAreaView>
        );
    }
}

const RootStack = createStackNavigator(
    {
        BuildingFootprintValidator,
        ChangeDetectionScreen,
        WelcomeScreen,
        ProjectNav,
        ProjectView,
        Mapper,
        Login,
        WebviewWindow,
    },
    {
        initialRouteName: 'WelcomeScreen',
        headerMode: 'none',
        navigationOptions: {
            gesturesEnabled: false,
        },
    },
);

module.exports = Main;
