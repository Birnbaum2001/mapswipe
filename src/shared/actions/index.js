// @flow

import { actionTypes } from 'react-redux-firebase';
import type { ResultMapType, ResultType, State } from '../flow-types';

export const WELCOME_COMPLETED: 'WELCOME_COMPLETED' = 'WELCOME_COMPLETED';
export const AUTH_STATUS_AVAILABLE: 'AUTH_STATUS_AVAILABLE' = 'AUTH_STATUS_AVAILABLE';
export const REQUEST_PROJECTS = 'REQUEST_PROJECTS';
export const TOGGLE_MAP_TILE: 'TOGGLE_MAP_TILE' = 'TOGGLE_MAP_TILE';
export const SUBMIT_BUILDING_FOOTPRINT = 'SUBMIT_BUILDING_FOOTPRINT';
export const COMMIT_GROUP = 'COMMIT_GROUP';
export const COMMIT_GROUP_FAILED = 'COMMIT_GROUP_FAILED';
export const COMMIT_GROUP_SUCCESS = 'COMMIT_GROUP_SUCCESS';
export const COMMIT_TASK_FAILED = 'COMMIT_TASK_FAILED';
export const COMMIT_TASK_SUCCESS = 'COMMIT_TASK_SUCCESS';

type CompleteWelcome = { type: typeof WELCOME_COMPLETED };
export function completeWelcome(): CompleteWelcome {
    return { type: WELCOME_COMPLETED };
}

type AuthStatusAvailable = { type: typeof AUTH_STATUS_AVAILABLE, user?: {}};
export function authStatusAvailable(user: {}): AuthStatusAvailable {
    return { type: AUTH_STATUS_AVAILABLE, user };
}

type ToggleMapTile = { type: typeof TOGGLE_MAP_TILE, tileInfo: ResultType };
export function toggleMapTile(tileInfo: ResultType): ToggleMapTile {
    // dispatched every time a map tile is tapped to change its state
    return { type: TOGGLE_MAP_TILE, tileInfo };
}

export function commitGroupSuccess(taskId: string) {
    return { type: COMMIT_GROUP_SUCCESS, taskId };
}

export function commitGroupFailed(taskId: string, error: {}) {
    return { type: COMMIT_GROUP_FAILED, taskId, error };
}

type CommitTaskSuccess = { type: typeof COMMIT_TASK_SUCCESS, taskId: number };
export function commitTaskSuccess(taskId: string) {
    return { type: COMMIT_TASK_SUCCESS, taskId };
}

type CommitTaskFailed = { type: typeof COMMIT_TASK_FAILED, taskId: number };
export function commitTaskFailed(taskId: string, error: {}) {
    return { type: COMMIT_TASK_FAILED, taskId, error };
}

type SubmitFootprint = { type: typeof SUBMIT_BUILDING_FOOTPRINT, resultObject: ResultType };
export function submitFootprint(resultObject: ResultType): SubmitFootprint {
    return { type: SUBMIT_BUILDING_FOOTPRINT, resultObject };
}

export type GroupInfo = {
    addedDistance: number,
    groupId: number,
    projectId: number,
    contributionsCount: number,
    results: ResultMapType,
    zoomLevel: number
}

type CommitGroup = { type: typeof COMMIT_GROUP }
export type Action =
    | actionTypes.SET_PROFILE
    | AuthStatusAvailable
    | CommitGroup
    | CommitTaskSuccess
    | CommitTaskFailed
    | CompleteWelcome
    | ToggleMapTile

type PromiseAction = Promise<Action>;
type GetState = () => State;
type GetFirebase = () => Object;
type Dispatch = (action: Action | PromiseAction) => any;
type ThunkAction = (dispatch: Dispatch, getState: GetState, getFirebase: GetFirebase) => any;

export function commitGroup(groupInfo: GroupInfo): ThunkAction {
    // dispatched when a group is finished, when the user chooses to either
    // map another group, or complete mapping.
    // Note that there are situations were the redux state tree will contain results from
    // a previously mapped group/project, so we are going to upload results for everything now.
    return (dispatch: Dispatch, getState: GetState, getFirebase: GetFirebase) => {
        const firebase = getFirebase();
        const userId = firebase.auth().currentUser.uid;

        Object.keys(groupInfo.results).forEach((resultId) => {
            const resObj = groupInfo.results[resultId];
            const fbpath = `results/${resObj.projectId}/${resObj.groupId}/${userId}/`;
            // this overwrites the timestamp if there are multiple results for a group
            firebase.set(`${fbpath}timestamp`, resObj.timestamp);
            firebase
                .set(`${fbpath}results/${resObj.resultId}`, resObj.result)
                .then(() => dispatch(commitTaskSuccess(resultId)))
                .catch(error => dispatch(commitTaskFailed(resultId, error)));
        });

        // increase the completion count on the group so we know how many users
        // have swiped through it
        // FIXME: chain all the promises above so that we can throw an action
        // once they have all completed
        firebase.database().ref(`groups/${groupInfo.projectId}/${groupInfo.groupId}`)
            .transaction((group) => {
                const newGroup = group;
                if (group) {
                    newGroup.completedCount += 1;
                }
                return newGroup;
            });

        // update the user's contributions and total mapped area
        const addedContributions = groupInfo.contributionsCount || 0;
        const addedDistance = groupInfo.addedDistance || 0;
        firebase.database().ref(`users/${userId}/`)
            .transaction((user) => {
                const newUser = user;
                if (user) {
                    newUser.contributions += addedContributions;
                    newUser.distance += addedDistance;
                }
                return newUser;
            });
    };
}
