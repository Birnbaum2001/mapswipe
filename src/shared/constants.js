// Constants for the project
// @flow

// project types
export const LEGACY_TILES: number = 1;
export const BUILDING_FOOTPRINTS: number = 2;
export const CHANGE_DETECTION: number = 3;
