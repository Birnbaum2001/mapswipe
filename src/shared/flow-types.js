// @flow
// type definitions. This should be safe to import everywhere
import { NavigationScreenProp, NavigationState } from 'react-navigation';
import BottomProgress from './views/Mapper/BottomProgress';

// geographic types

export type Point = [number, number];

export type Polygon = Array<Point>;

export type BBOX = [number, number, number, number];

export type Tile = [number, number, number];

// dependencies types

export type NavigationProp = NavigationScreenProp<NavigationState>;

// types related to firebase data

export type ProjectType = {
    contributors: number,
    image: string,
    info: { apiKey: string, tileServerUrl: string },
    isFeatured: boolean,
    lookFor: string,
    name: string,
    projectDetails: string,
    projectId: number,
    projectType: ?number,
    progress: number,
    state: number,
};

export type ProjectMapType = { [project_id: string]: ProjectType };

export type TaskType = {
    groupId: number,
    featureId?: number,
    geojson?: { type: string, coordinates: { [number]: Polygon }},
    projectId: number,
    taskId: string,
    taskX: number,
    taskY: number,
    url: string,
}

export type TaskMapType = { [task_id: string]: TaskType };

export type GroupType = {
    groupId: number,
    neededCount: number,
    projectId: number,
    tasks: Array<TaskType>,
    zoomLevel: number,
    xMax: number;
    xMin: number;
    yMax: number;
    yMin: number;
}

export type GroupMapType = { [group_id: string]: GroupType };

export type ResultType = {
    groupId: number,
    projectId: number,
    resultId: string,
    result: number,
    timestamp: {},
}

export type ResultMapType = { [string]: ResultType };

// internal app types

export type Mapper = {
    closeTilePopup: () => void,
    openTilePopup: (any) => void,
    progress: BottomProgress,
    project: ProjectType,
}

// redux types

export type UIState = {
    welcomeCompleted: boolean,
};

export type State = {
    +firebase?: {},
    +results?: ResultMapType,
    +ui?: UIState,
};
