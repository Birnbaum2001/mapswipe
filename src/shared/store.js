// @flow
import { Platform } from 'react-native';
import { applyMiddleware, createStore } from 'redux';
import thunkMiddleware from 'redux-thunk';
import firebase from 'react-native-firebase';
import { getFirebase, reactReduxFirebase } from 'react-redux-firebase';
import { composeWithDevTools } from 'remote-redux-devtools';
import reducers from './reducers/index';

const reactFirebaseConfig = {
    attachAuthIsReady: true,
    enableRedirectHandling: false,
    userProfile: 'users',
};

const composeEnhancers = composeWithDevTools({
    name: Platform.OS,
    hostname: 'localhost',
    port: 5678,
    realtime: false,
});

export const store = createStore(
    reducers,
    composeEnhancers(
        applyMiddleware(thunkMiddleware.withExtraArgument(getFirebase)),
        reactReduxFirebase(firebase, reactFirebaseConfig),
    ),
);

export default function setupStore() {
    return store;
}
