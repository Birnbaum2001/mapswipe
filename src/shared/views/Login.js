/**
 * Created by wwadewitte on 7/11/16.
 */

// @flow

import React from 'react';
import { compose } from 'redux';
import { connect } from 'react-redux';
import { firebaseConnect, isEmpty, isLoaded } from 'react-redux-firebase';
import { withNamespaces } from 'react-i18next';
import {
    Text,
    View,
    ScrollView,
    StyleSheet,
    Image,
    TextInput,
} from 'react-native';
import Button from 'apsl-react-native-button';
import { MessageBarManager } from 'react-native-message-bar';
import LoadingIcon from './LoadingIcon';
import { store as reduxStore } from '../store';
import type { NavigationProp } from '../flow-types';

/* eslint-disable global-require */

const GLOBAL = require('../Globals');

const styles = StyleSheet.create({
    switchToLogin: {
        width: GLOBAL.SCREEN_WIDTH * 0.90,
        height: 50,
        padding: 12,
        borderRadius: 0,
        borderWidth: 0,
    },
    otherButton: {
        backgroundColor: 'rgb(255,25,25)',
        width: GLOBAL.SCREEN_WIDTH * 0.90,
        height: 50,
        padding: 12,
        borderRadius: 5,
        borderWidth: 0.1,
    },
    container: {
        width: GLOBAL.SCREEN_WIDTH,
        height: GLOBAL.SCREEN_HEIGHT,
        backgroundColor: '#0d1949',
        flex: 1,
    },
    text4: {
        width: GLOBAL.SCREEN_WIDTH * 0.90,
        marginTop: 10,
        textAlign: 'left',
        fontSize: 13,
        marginBottom: 10,
        color: '#ffffff',
    },
    text5: {
        width: GLOBAL.SCREEN_WIDTH * 0.90,
        marginTop: 10,
        textAlign: 'left',
        fontSize: 10,
        marginBottom: 10,
        color: '#ffffff',
    },
    tutIcon2: {
        resizeMode: 'contain',
        width: 100,
        height: 100,
        marginBottom: 30,
        marginTop: 30,
    },
    textInput: {
        width: GLOBAL.SCREEN_WIDTH * 0.90,
        height: 40,
        backgroundColor: 'rgba(255,255,255,0.2)',
        borderRadius: 5,
        color: '#ffffff',
        paddingLeft: 10,
    },
});

// Screen constants
const SCREEN_SIGNUP = 0;
const SCREEN_LOGIN = 1;
const SCREEN_FORGOT_PASSWORD = 2;

type Props = {
    auth: {},
    firebase: Object,
    navigation: NavigationProp,
    t: (string) => string,
}

type State = {
    username: string,
    password: string,
    email: string,
    loading: boolean,
    screen: number
}

class _Login extends React.Component<Props, State> {
    constructor(props: Props) {
        super(props);
        this.state = {
            username: '',
            password: '',
            email: '',
            loading: true,
            screen: SCREEN_SIGNUP,
        };
        const that = this;
        reduxStore.firebaseAuthIsReady.then(() => that.setState({ loading: false }));
    }

    componentDidMount() {
        const { auth, navigation } = this.props;
        if (isLoaded(auth) && !isEmpty(auth)) {
            navigation.push('ProjectNav');
        }
    }

    componentDidUpdate(prevProps: Props) {
        const { auth, navigation } = this.props;
        if (auth !== prevProps.auth) {
            if (isLoaded(auth) && !isEmpty(auth)) {
                navigation.push('ProjectNav');
            }
        }
    }

    handleSignUp = () => {
        // GLOBAL.ANALYTICS.logEvent('account_screen_seen');
        const {
            firebase,
            navigation,
        } = this.props;
        const {
            email,
            password,
            username,
        } = this.state;
        const parent = this;
        if (username !== null && username.length < 3) {
            // GLOBAL.ANALYTICS.logEvent('account_creation_error_username');
            MessageBarManager.showAlert({
                title: 'Error on sign up',
                message: 'Your username must be 4 characters or more',
                alertType: 'error',
            });
            return;
        }

        if (username !== null && username.indexOf('@') !== -1) {
            // GLOBAL.ANALYTICS.logEvent('account_creation_error_username');
            MessageBarManager.showAlert({
                title: 'Error on sign up',
                message: 'Your username can not be an email',
                alertType: 'error',
            });
            return;
        }
        this.setState({
            loading: true,
        });

        firebase.createUser({ email, password }, { username })
            .then(() => {
                // firebase doesn't record the username by default
                firebase.updateAuth({
                    displayName: username,
                });
            })
            .then(() => {
                firebase.reloadAuth();
            })
            .then(() => {
                firebase.updateProfile({
                    distance: 0,
                    contributions: 0,
                });
            })
            .then(() => {
                MessageBarManager.showAlert({
                    title: 'Success',
                    message: `Welcome to Mapswipe, ${username}`,
                    alertType: 'info',
                });
                parent.setState({
                    loading: false,
                });
                // GLOBAL.ANALYTICS.logEvent('account_created');
                navigation.push('ProjectNav');
            })
            .catch((error) => {
                // GLOBAL.ANALYTICS.logEvent('account_creation_error_db');
                let errorMsg;
                // error codes from https://rnfirebase.io/docs/v5.x.x/auth/reference/auth#createUserWithEmailAndPassword
                switch (error.code) {
                case 'auth/email-already-in-use':
                    errorMsg = 'Email already used by another account';
                    break;
                case 'auth/invalid-email':
                    errorMsg = 'Email address is invalid';
                    break;
                default:
                    errorMsg = 'Problem signing up';
                }
                MessageBarManager.showAlert({
                    title: 'Error on sign up',
                    message: errorMsg,
                    alertType: 'error',
                });
                parent.setState({
                    loading: false,
                });
            });
    }

    switchScreens = (screen: number) => {
        this.setState({
            screen,
        });
    }

    handleLogin = () => {
        const {
            email,
            password,
        } = this.state;
        const { firebase } = this.props;
        this.setState({
            loading: true,
        });
        const parent = this;
        firebase.login({ email, password }).then((userCredentials) => {
            MessageBarManager.showAlert({
                title: 'Success',
                message: `Welcome to Mapswipe, ${userCredentials.user.user.displayName}`,
                alertType: 'info',
            });
            // GLOBAL.ANALYTICS.logEvent('account_login');
            parent.props.navigation.push('ProjectNav');
        }).catch((error) => {
            let errorMessage;
            // error codes from
            // https://rnfirebase.io/docs/v5.x.x/auth/reference/auth#signInWithEmailAndPassword
            switch (error.code) {
            case 'auth/user-not-found':
                errorMessage = 'No account with this email address';
                break;
            case 'auth/wrong-password':
            case 'auth/invalid-email':
                errorMessage = 'Invalid email or password';
                break;
            case 'auth/user-disabled':
                errorMessage = 'Account disabled';
                break;
            default:
                errorMessage = 'Problem logging in';
            }
            MessageBarManager.showAlert({
                title: 'Error on log in',
                message: errorMessage,
                alertType: 'error',
            });
            parent.setState({
                loading: false,
            });
        });
    }

    handlePassReset() {
        const {
            email,
        } = this.state;
        const { firebase } = this.props;
        this.setState({
            loading: true,
        });
        const parent = this;
        firebase.sendPasswordResetEmail(email).then(() => {
            MessageBarManager.showAlert({
                title: 'Success',
                message: 'Check your email',
                alertType: 'info',
            });
            parent.setState({
                loading: false,
            });
            // GLOBAL.ANALYTICS.logEvent('pass_reset_request');
        }).catch((error) => {
            let errorMessage;
            switch (error.code) {
            case 'auth/user-not-found':
                errorMessage = 'No account found for this email';
                break;
            case 'auth/invalid-email':
                errorMessage = 'Invalid email address';
                break;
            default:
                errorMessage = 'Problem resetting your password';
            }
            MessageBarManager.showAlert({
                title: 'Error on reset pass',
                message: errorMessage,
                alertType: 'error',
            });
            parent.setState({
                loading: false,
            });
        });
    }

    renderSignupScreen = () => {
        const { t } = this.props;
        const {
            email,
            username,
        } = this.state;
        return (
            <ScrollView
                style={styles.container}
                contentContainerStyle={{
                    alignItems: 'center',
                    width: GLOBAL.SCREEN_WIDTH,
                    backgroundColor: '#0d1949',
                    padding: 20,
                }}
            >
                <Image style={styles.tutIcon2} source={require('./assets/loadinganimation.gif')} />
                <Text style={styles.text4}>{t('login:enterUsername')}</Text>

                <TextInput
                    autoCorrect={false}
                    style={styles.textInput}
                    onChangeText={text => this.setState({ username: text })}
                    value={username}
                />

                <Text style={styles.text4}>Enter your email</Text>

                <TextInput
                    autoCorrect={false}
                    secureTextEntry={false}
                    style={styles.textInput}
                    onChangeText={text => this.setState({ email: text.replace(' ', '') })}
                    value={email}
                />
                <Text style={styles.text4}>Enter your password (More than 6 characters)</Text>

                <TextInput
                    autoCorrect={false}
                    secureTextEntry
                    style={styles.textInput}
                    onChangeText={text => this.setState({ password: text })}
                />
                <Text style={styles.text5}>
                    * All the mapping you contribute to mapswipe is open and available to anyone.
                    Your username is public, but your email and password will never be
                    shared with anyone.
                </Text>
                <Button
                    style={styles.otherButton}
                    onPress={this.handleSignUp}
                    textStyle={{ fontSize: 13, color: '#ffffff', fontWeight: '700' }}
                >
                    Sign Up
                </Button>
                <Button
                    style={styles.switchToLogin}
                    onPress={() => this.switchScreens(SCREEN_LOGIN)}
                    textStyle={{ fontSize: 13, color: '#ffffff', fontWeight: '700' }}
                >
                    Log in to an existing account
                </Button>
            </ScrollView>
        );
    }

    renderLoginScreen = () => {
        const {
            email,
        } = this.state;
        return (
            <ScrollView
                style={styles.container}
                contentContainerStyle={{
                    alignItems: 'center',
                    width: GLOBAL.SCREEN_WIDTH,
                    backgroundColor: '#0d1949',
                    padding: 20,
                }}
            >
                <Image style={styles.tutIcon2} source={require('./assets/loadinganimation.gif')} />

                <Text style={styles.text4}>Enter your email</Text>
                <TextInput
                    autoCorrect={false}
                    style={styles.textInput}
                    secureTextEntry={false}
                    onChangeText={text => this.setState({ email: text.replace(' ', '') })}
                    value={email}
                />
                <Text style={styles.text4}>Enter your password</Text>

                <TextInput
                    autoCorrect={false}
                    secureTextEntry
                    style={styles.textInput}
                    onChangeText={text => this.setState({ password: text })}
                />
                <Text style={styles.text5}>
                    * All the data you contribute to mapswipe is open and available to anyone.
                    Your username is public, but your email and password will never be
                    shared with anyone.
                    {' '}
                </Text>
                <Button
                    style={styles.otherButton}
                    onPress={this.handleLogin}
                    textStyle={{ fontSize: 13, color: '#ffffff', fontWeight: '700' }}
                >
                    Log in
                </Button>
                <Button
                    style={styles.switchToLogin}
                    onPress={() => this.switchScreens(SCREEN_FORGOT_PASSWORD)}
                    textStyle={{ fontSize: 13, color: '#ffffff', fontWeight: '700' }}
                >
                    Forgot your password?
                </Button>
                <Button
                    style={styles.switchToLogin}
                    onPress={() => this.switchScreens(SCREEN_SIGNUP)}
                    textStyle={{ fontSize: 13, color: '#ffffff', fontWeight: '700' }}
                >
                    Create New Account
                </Button>
            </ScrollView>
        );
    }

    renderForgotPasswordScreen = () => {
        const {
            email,
        } = this.state;
        return (
            <ScrollView
                style={styles.container}
                contentContainerStyle={{
                    alignItems: 'center',
                    width: GLOBAL.SCREEN_WIDTH,
                    backgroundColor: '#0d1949',
                    padding: 20,
                }}
            >
                <Image style={styles.tutIcon2} source={require('./assets/loadinganimation.gif')} />

                <Text style={styles.text4}>Enter your email</Text>
                <TextInput
                    autoCorrect={false}
                    style={styles.textInput}
                    onChangeText={text => this.setState({ email: text.replace(' ', '') })}
                    value={email}
                />
                <Text style={styles.text5}>* We will send you an email to reset your password</Text>
                <Button
                    style={styles.otherButton}
                    onPress={this.handlePassReset}
                    textStyle={{ fontSize: 13, color: '#ffffff', fontWeight: '700' }}
                >
                    Send Reset Email
                </Button>
                <Button
                    style={styles.switchToLogin}
                    onPress={() => this.switchScreens(SCREEN_LOGIN)}
                    textStyle={{ fontSize: 13, color: '#ffffff', fontWeight: '700' }}
                >
                    Back to login
                </Button>
            </ScrollView>
        );
    }

    render() {
        const { auth } = this.props;
        const {
            loading,
            screen,
        } = this.state;
        let content;

        if (screen === SCREEN_SIGNUP) {
            content = this.renderSignupScreen();
        } else if (screen === SCREEN_LOGIN) {
            content = this.renderLoginScreen();
        } else if (screen === SCREEN_FORGOT_PASSWORD) {
            content = this.renderForgotPasswordScreen();
        }

        return (
            <View style={styles.container}>
                {
                    !isLoaded(auth) || loading
                        ? <LoadingIcon />
                        : content
                }
            </View>
        );
    }
}

const mapStateToProps = (state, ownProps) => (
    {
        auth: state.firebase.auth,
        navigation: ownProps.navigation,
        profile: state.firebase.profile,
    }
);

const enhance = compose(
    withNamespaces('login'),
    firebaseConnect(),
    connect(
        mapStateToProps,
    ),
);

export default enhance(_Login);
