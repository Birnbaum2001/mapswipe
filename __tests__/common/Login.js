import React from 'react';
import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';
import Login from '../../src/shared/views/Login';
import { store } from '../../src/shared/store';

test('Renders sign up screen', () => {
    // this only renders the loading icon, as auth is not ready yet
    const l = <Provider store={store}><Login /></Provider>;
    const screen = renderer.create(l).toJSON();
    expect(screen).toMatchSnapshot();
});
