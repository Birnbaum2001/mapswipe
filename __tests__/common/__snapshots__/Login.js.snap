// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders sign up screen 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#0d1949",
      "flex": 1,
      "height": 1334,
      "width": 750,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "flex": 1,
        "height": 1264,
        "justifyContent": "center",
        "opacity": 0,
        "width": 750,
      }
    }
  >
    <Image
      source={
        Object {
          "testUri": "../../../src/shared/views/assets/loadinganimation.gif",
        }
      }
      style={
        Object {
          "height": 100,
          "width": 100,
        }
      }
    />
    <Text
      style={
        Object {
          "color": "#ffffff",
          "fontSize": 20,
          "fontWeight": "300",
          "marginTop": 20,
        }
      }
    >
      Loading...
    </Text>
  </View>
</View>
`;
